src/v/cluster_link/replication/partition_replicator.cc

```diff
@@ -0,0 +1,129 @@
+
+partition_replicator::partition_replicator(
+  const model::ntp& ntp,
+  model::term_id term,
+  std::unique_ptr<data_source> source,
+  std::unique_ptr<data_sink> sink)
+  : _term(term)
+  , _log(cllog, fmt::format("[{}-term-{}] replicator", ntp, term))
+  , _source(std::move(source))
+  , _sink(std::move(sink)) {}
+
+ss::future<> partition_replicator::start() {
+    vlog(_log.trace, "Starting replicator");
+    co_await _source->start();
+    co_await _sink->start();
+    ssx::repeat_until_gate_closed(_gate, [this] {
+        return fetch_and_replicate().handle_exception(
+          [this](const std::exception_ptr& e) {
+              auto log_level = ssx::is_shutdown_exception(e)
+                                 ? ss::log_level::trace
+                                 : ss::log_level::warn;
+              _log.log(log_level, "Error in partition replicator: {}", e);
+          });
+    });
+}
+
+ss::future<> partition_replicator::stop() {
+    vlog(_log.trace, "Stopping replicator");
+    _as.request_abort();
+    auto f = _gate.close();
+    co_await _source->stop();
+    co_await _sink->stop();
+    co_await std::move(f);
+    vlog(_log.trace, "Stopped replicator");
+}
+
+ss::future<> partition_replicator::wait_for_replication_result(
+  ss::future<result<raft::replicate_result>> f,
+  model::offset begin,
+  model::offset end,
+  ss::abort_source& as) {
+    try {
+        auto result = co_await std::move(f);
+        if (result.has_error()) {
+            vlog(
+              _log.warn,
+              "Replication of batches in range [{} - {}] failed with error: {}",
+              begin,
+              end,
+              result.error());
+            as.request_abort();
+            co_return;
+        }
+        vlog(
+          _log.trace,
+          "Replicated batches in range [{} - {}] with at offset: {}",
+          begin,
+          end,
+          result.value().last_offset);
+    } catch (...) {
+        vlog(
+          _log.error,
+          "Exception during replication: {}",
+          std::current_exception());
+        as.request_abort();
+    }
+}
+
+ss::future<> partition_replicator::fetch_and_replicate() {
+    _gate.check();
+    // abort source for this iteration of fetch_and_replicate
+    ss::abort_source as;
+    auto subscription = _as.subscribe([&as] noexcept { as.request_abort(); });
+    co_await _source->reset(
+      kafka::next_offset(_sink->last_replicated_offset()));
+    ss::gate gate;
+    try {
+        while (!_gate.is_closed() && !as.abort_requested()) {
+            auto inflight_units = co_await ss::get_units(_max_requests, 1, as);
+            auto data = co_await _source->fetch_next(as);
+            if (data.batches.empty()) {
+                continue;
+            }
+            auto begin = data.batches.front().base_offset();
+            auto end = data.batches.back().last_offset();
+            as.check();
+            auto stages = co_await _sink->replicate(std::move(data.batches));
```

Comment:
i am thinking if it would be good to pass an abort source into `replicate` this way we have a control over waiting for replication result.